// connect-architect/connect.formauth/formauth.js
"use strict";

var crypto = require("crypto");
var cookie = require("cookie");
var signature = require("cookie-signature");

var COOKIE_NAME = "c9auth";
var COOKIE_MAX_AGE = 86400; // 24 hours in seconds

// IP-based rate limiting: max 10 attempts per 15 minutes
var RATE_WINDOW_MS = 15 * 60 * 1000;
var RATE_MAX = 10;
var rateLimitMap = {};

function getClientIP(req) {
    var xff = req.headers["x-forwarded-for"];
    if (xff) return xff.split(",")[0].trim();
    return (req.socket && req.socket.remoteAddress) || "unknown";
}

function isRateLimited(ip) {
    var now = Date.now();
    var entry = rateLimitMap[ip];
    if (!entry || now - entry.start > RATE_WINDOW_MS) {
        rateLimitMap[ip] = { start: now, count: 1 };
        return false;
    }
    entry.count++;
    return entry.count > RATE_MAX;
}

function verifyPassword(plain, storedHash, cb) {
    // Format: "iterations:saltHex:hashHex"
    var parts = storedHash.split(":");
    if (parts.length !== 3) return cb(null, false);
    var iters = parseInt(parts[0], 10);
    var salt = Buffer.from(parts[1], "hex");
    var expected = Buffer.from(parts[2], "hex");
    crypto.pbkdf2(plain, salt, iters, 64, "sha512", function(err, actual) {
        if (err) return cb(null, false);
        try {
            cb(null, crypto.timingSafeEqual(actual, expected));
        } catch (e) {
            cb(null, false);
        }
    });
}

function makeSessionToken(secret) {
    var rand = crypto.randomBytes(32).toString("hex");
    return signature.sign(rand, secret);
}

function validateSessionToken(token, secret) {
    if (!token) return false;
    return !!signature.unsign(token, secret);
}

function loginPage(error) {
    var errHtml = error
        ? '<p style="color:#c0392b;margin-bottom:12px">' + escapeHtml(error) + "</p>"
        : "";
    return "<!DOCTYPE html><html><head><meta charset=\"utf-8\">" +
        "<title>Cloud9 Login</title>" +
        "<style>body{font-family:sans-serif;display:flex;justify-content:center;" +
        "align-items:center;height:100vh;margin:0;background:#1e1e1e}" +
        ".box{background:#2d2d2d;padding:2rem 2.5rem;border-radius:8px;width:300px}" +
        "h2{color:#e0e0e0;margin:0 0 1.5rem}label{color:#aaa;font-size:.85rem}" +
        "input{width:100%;padding:.5rem;margin:.25rem 0 1rem;box-sizing:border-box;" +
        "background:#3a3a3a;border:1px solid #555;border-radius:4px;color:#e0e0e0}" +
        "button{width:100%;padding:.6rem;background:#2980b9;color:#fff;border:none;" +
        "border-radius:4px;cursor:pointer;font-size:1rem}" +
        "button:hover{background:#3498db}</style></head>" +
        "<body><div class=\"box\"><h2>Cloud9</h2>" + errHtml +
        "<form method=\"POST\" action=\"/login\">" +
        "<label>Username</label><input name=\"username\" autofocus autocomplete=\"username\">" +
        "<label>Password</label><input name=\"password\" type=\"password\" autocomplete=\"current-password\">" +
        "<button type=\"submit\">Sign in</button></form></div></body></html>";
}

function escapeHtml(str) {
    return String(str)
        .replace(/&/g, "&amp;")
        .replace(/</g, "&lt;")
        .replace(/>/g, "&gt;")
        .replace(/"/g, "&quot;");
}


module.exports = function setup(options, imports, register) {
    var connect = imports.connect;
    var username = options.username || "";
    var passwordHash = options.passwordHash || "";
    var secure = options.secure || false;

    // If no password configured, skip auth entirely
    if (!passwordHash) {
        register(null, {});
        return;
    }

    // Use a per-process secret for session tokens — sessions do not survive restarts
    var secret = crypto.randomBytes(64).toString("hex");

    connect.use(function formauth(req, res, next) {
        var url = req.url.split("?")[0];

        // Serve login page
        if (url === "/login" && req.method === "GET") {
            res.writeHead(200, { "Content-Type": "text/html; charset=utf-8" });
            return res.end(loginPage());
        }

        // Handle login form submission
        // Note: connect.urlencoded() runs before plugins, so req.body is already parsed
        if (url === "/login" && req.method === "POST") {
            var ip = getClientIP(req);
            if (isRateLimited(ip)) {
                res.writeHead(429, { "Content-Type": "text/html; charset=utf-8" });
                return res.end(loginPage("Too many attempts. Try again in 15 minutes."));
            }

            var body = req.body || {};
            verifyPassword(body.password || "", passwordHash, function(err, ok) {
                if (!ok || body.username !== username) {
                    res.writeHead(401, { "Content-Type": "text/html; charset=utf-8" });
                    return res.end(loginPage("Invalid username or password."));
                }

                var token = makeSessionToken(secret);
                var cookieOpts = {
                    httpOnly: true,
                    path: "/",
                    maxAge: COOKIE_MAX_AGE,
                    sameSite: "Lax"
                };
                if (secure) cookieOpts.secure = true;

                var redirectTo = "/";
                var referer = req.headers["referer"] || "";
                var match = referer.match(/[?&]redirect=([^&]*)/);
                if (match) {
                    var dest = decodeURIComponent(match[1]);
                    // Prevent open redirect: only allow relative paths
                    if (/^\/[^/\\]/.test(dest) || dest === "/") redirectTo = dest;
                }

                res.writeHead(302, {
                    "Set-Cookie": cookie.serialize(COOKIE_NAME, token, cookieOpts),
                    "Location": redirectTo
                });
                res.end();
            });
            return;
        }

        // Handle logout
        if (url === "/logout") {
            res.writeHead(302, {
                "Set-Cookie": cookie.serialize(COOKIE_NAME, "", { httpOnly: true, path: "/", maxAge: 0 }),
                "Location": "/login"
            });
            return res.end();
        }

        // Validate session cookie on all other requests
        var cookies = cookie.parse(req.headers.cookie || "");
        var token = cookies[COOKIE_NAME];
        if (validateSessionToken(token, secret)) {
            return next();
        }

        // Not authenticated — redirect to login
        var redirectParam = encodeURIComponent(req.url);
        res.writeHead(302, { "Location": "/login?redirect=" + redirectParam });
        res.end();
    });

    register(null, {});
};
